internal/crypto/cipher.go

```diff
@@ -0,0 +1,208 @@
+package crypto
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"fmt"
+	"hash"
+
+	"github.com/optable/match/internal/util"
+	"github.com/zeebo/blake3"
+)
+
+/*
+Various cipher suite implementation in golang
+*/
+
+const (
+	GCM = iota
+	XORBlake3
+
+	nonceSize = 12 //aesgcm NonceSize
+)
+
+// PseudorandomCode is implemented as follows:
+// C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
+// secretKey is a 16 byte slice for AES-128
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCode(aesBlock cipher.Block, src []byte) (dst []byte) {
+	dst = make([]byte, aes.BlockSize*4)
+	// effectively pad src
+	copy(dst, src)
+
+	// encrypt
+	aesBlock.Encrypt(dst[:aes.BlockSize], dst[:aes.BlockSize])
+	if len(src) <= aes.BlockSize {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize:aes.BlockSize*2], dst[aes.BlockSize:aes.BlockSize*2])
+	if len(src) <= aes.BlockSize*2 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*2:aes.BlockSize*3], dst[aes.BlockSize*2:aes.BlockSize*3])
+	if len(src) <= aes.BlockSize*3 {
+		return dst
+	}
+
+	aesBlock.Encrypt(dst[aes.BlockSize*3:], dst[aes.BlockSize*3:])
+	return dst
+}
+
+// PseudorandomCodeWithHashIndex is implemented as follows:
+// C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
+// secretKey is a 16 byte slice for AES-128
+// on success, pseudorandomCode returns a byte slice of 64 bytes.
+func PseudorandomCodeWithHashIndex(aesBlock cipher.Block, src []byte, hIdx byte) (dst []byte) {
+	dst = make([]byte, aes.BlockSize*4)
+	// effectively pad src
+	copy(dst, src)
+	if len(src) < aes.BlockSize*4 {
+		dst[len(src)] = hIdx
```

Comment:
If src contains 4 blocks, hidx is ignored?

Reply:
Yes, appending the hash index is only to reduce the risk of collision. For src that already contain 4 blocks, there is no reason to append the hash index as it won't be used at all. Also that append, was quite costly for memory and performance.

Reply:
I'm wondering if we could share the block encryption part in a common subroutine that takes a number of blocks to encrypt in place in a 64 byte slice and keep the padding and hashIdx appending in the public functions.

Reply:
Something like this? It is slightly slower than the current implementation.
```
func PseudorandomCodeWithHashIndex(aesBlock cipher.Block, src []byte, hIdx byte) (dst []byte) {
	dst = make([]byte, aes.BlockSize*4)
	// effectively pad src
	copy(dst, src)
	if len(src) < aes.BlockSize*4 {
		dst[len(src)] = hIdx
	}

	var blocks int
	switch true {
	case len(src)+1 > aes.BlockSize*3:
		blocks = 4
	case len(src)+1 > aes.BlockSize*2:
		blocks = 3
	case len(src)+1 > aes.BlockSize*1:
		blocks = 2
	default:
		blocks = 1
	}

	// encrypt
	aesEncrypt(aesBlock, blocks, dst, dst)
	return dst
}

func aesEncrypt(aesBlock cipher.Block, blocks int, src, dst []byte) {
	for b := 0; b < blocks; b++ {
		aesBlock.Encrypt(dst[aes.BlockSize*b:aes.BlockSize*(b+1)], dst[aes.BlockSize*b:aes.BlockSize*(b+1)])
	}
}
```

Reply:
Okay, modified like this to avoid switch case:
```
func PseudorandomCodeWithHashIndex(aesBlock cipher.Block, src []byte, hIdx byte) (dst []byte) {
	dst = make([]byte, aes.BlockSize*4)
	// effectively pad src
	copy(dst, src)
	if len(src) < aes.BlockSize*4 {
		dst[len(src)] = hIdx
	}

	// encrypt
	aesEncrypt(aesBlock, (len(src)/aes.BlockSize)+1, dst)
	return dst
}

func aesEncrypt(aesBlock cipher.Block, blocks int, src []byte) {
	for b := 0; b < blocks && b < 4; b++ {
		aesBlock.Encrypt(src[aes.BlockSize*b:aes.BlockSize*(b+1)], src[aes.BlockSize*b:aes.BlockSize*(b+1)])
	}
}
```

Reply:
This modified version very slightly underperforms:
```
BenchmarkPseudorandomCodeWithHashIndex
BenchmarkPseudorandomCodeWithHashIndex-12       17904160                62.44 ns/op           64 B/op          1 allocs/op
BenchmarkPseudorandomCodeWithHashIndex2
BenchmarkPseudorandomCodeWithHashIndex2-12      16709091                68.98 ns/op           64 B/op          1 allocs/op
```